File: src/main/java/ru/albemuth/analysis/population/Woman.java
package ru.albemuth.analysis.population;

/**
 * @author VKornyshev
 * Класс, моделирующий женщину
 */
public class Woman extends Human {

    protected int laborCounter = 0;//кол-во родов

    public Woman(int ageLimit, double firstLaborDeathP) {
        super(ageLimit, firstLaborDeathP);
    }

    @Override
    public Sex getSex() {
        return Sex.FEMALE;
    }

    public int getLaborCounter() {
        return laborCounter;
    }

    public void incLaborCounter() {
        laborCounter++;
    }

    public boolean isLaborAccepted() {
        //женщина рожает на 17, 21, 25, 29, 33 год
        return age == 17 || age == 21 || age == 25 || age == 29 || age == 33;
    }

    public double calculateLaborDeathP(int laborCounter, double firstLaborDeathP) {
        double laborDeathP;
        switch (laborCounter) {
            case 1:
                laborDeathP = firstLaborDeathP;
                break;
            case 2:
                laborDeathP = 0.7 * firstLaborDeathP;
                break;
            case 3:
                laborDeathP = 0.5 * firstLaborDeathP;
                break;
            case 4:
                laborDeathP = 0.7 * firstLaborDeathP;
                break;
            case 5:
                laborDeathP = firstLaborDeathP;
                break;
            default:
                laborDeathP = 1;
        }
        return laborDeathP;
    }

}

File: src/main/scala/ru/albemuth/analysis/population/v3/Woman.scala
package ru.albemuth.analysis.population.v3

import ru.albemuth.analysis.population.v3.Human.Sex

/**
 * @author vovan
 * Класс, моделирующий женщину
 */
object Woman {

    def calculateLaborDeathP(d: Double): Double = {
        import scala.math.exp
        val r = 16.606082245039158
        Population.p(0.001 * exp(r * d) / (1 + 0.001 * (exp(r * d) - 1)))
    }
    
    def calculateLaborDeathP(laborDeathP: Double, laborCounter: Int): Double = {
        var deathP: Double = 0

        laborCounter match {
            case 1 => deathP = laborDeathP
            case 2 => deathP = 0.7 * laborDeathP
            case 3 => deathP = 0.5 * laborDeathP
            case 4 => deathP = 0.7 * laborDeathP
            case 5 => deathP = laborDeathP
            case _ => deathP = laborDeathP
        }

        deathP
    }

}

class Woman(aCranialDiameter: Double, anAge: Int, anAgeLimit: Int) extends Human(aCranialDiameter, anAge, anAgeLimit) {

    val sex = Sex.FEMALE

    private var laborCounter_ = 0

    def laborCounter = laborCounter_
    def incLaborCounter() { laborCounter_ += 1}

    def isLaborAccepted: Boolean = {
        //женщина рожает на 17, 21, 25, 29, 33 год
        age == 17 || age == 21 || age == 25 || age == 29 || age == 33
    }

    def calculateLaborDeathP(childCranialDiameter: Double): Double = {Woman.calculateLaborDeathP(Woman.calculateLaborDeathP(childCranialDiameter), laborCounter)}

}
